File: executor/services/updater/updater_service.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""Updater service - orchestrates the executor self-update flow.

Coordinates version checking, downloading, and binary replacement
with clear user interaction and error handling.
"""

import logging
import os
import shutil
import sys
from dataclasses import dataclass
from logging.handlers import RotatingFileHandler
from pathlib import Path
from typing import Optional

from executor.config.device_config import UpdateConfig
from executor.services.updater.binary_replacer import BinaryReplacer
from executor.services.updater.version_checker_factory import create_version_checker

# Use 'updater' logger to write to upgrade.log
logger = logging.getLogger("updater")


def setup_upgrade_logging(verbose: bool = False) -> logging.Logger:
    """
    Setup upgrade logging. Only enable file logging when verbose is True.
    Always add console handler for basic user feedback.

    Args:
        verbose: If True, enable file logging to upgrade.log with DEBUG level.
                If False, only console logging is enabled with INFO level.

    Returns:
        Configured logger instance for updater.
    """
    logger = logging.getLogger("updater")

    if logger.handlers:
        return logger

    # Console handler - always enabled for basic feedback
    console = logging.StreamHandler(sys.stdout)
    console.setLevel(logging.INFO)
    console.setFormatter(logging.Formatter("%(message)s"))
    logger.addHandler(console)

    if verbose:
        # File handler - only when verbose
        log_dir = Path.home() / ".wegent-executor" / "logs"
        log_dir.mkdir(parents=True, exist_ok=True)
        file_handler = RotatingFileHandler(
            log_dir / "upgrade.log",
            maxBytes=10 * 1024 * 1024,
            backupCount=5,
            encoding="utf-8",
        )
        file_handler.setLevel(logging.DEBUG)
        file_handler.setFormatter(
            logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
        )
        logger.addHandler(file_handler)
        logger.setLevel(logging.DEBUG)
    else:
        logger.setLevel(logging.INFO)

    return logger


@dataclass
class UpdateResult:
    """Result of an update operation.

    Attributes:
        success: True if update completed successfully
        already_latest: True if already on the latest version
        old_version: Version before update (if applicable)
        new_version: Version after update (if applicable)
        error: Error message if update failed
    """

    success: bool = False
    already_latest: bool = False
    old_version: Optional[str] = None
    new_version: Optional[str] = None
    error: Optional[str] = None


class UpdaterService:
    """Orchestrate the executor self-update process.

    Coordinates version checking, user confirmation, downloading,
    and binary replacement with comprehensive error handling.
    """

    # Minimum free disk space required (100 MB with 50% safety margin = 150 MB)
    MIN_FREE_SPACE = 150 * 1024 * 1024

    def __init__(
        self,
        update_config: UpdateConfig,
        auto_confirm: bool = False,
        verbose: bool = False,
    ):
        """Initialize the updater service.

        Args:
            update_config: Update configuration from device config
            auto_confirm: If True, skip user confirmation prompts
            verbose: If True, enable verbose logging to file
        """
        self.update_config = update_config
        self.version_checker = create_version_checker(update_config)
        self.binary_replacer: Optional[BinaryReplacer] = None
        self.auto_confirm = auto_confirm

        # Setup logging (uses unified setup_upgrade_logging function)
        setup_upgrade_logging(verbose=verbose)

    async def check_and_update(self) -> UpdateResult:
        """Main update flow orchestration.

        Performs the complete update process:
        1. Get current version
        2. Check for updates via API
        3. If update available, prompt user for confirmation
        4. Download new binary with progress display
        5. Replace binary atomically with backup
        6. Return result with status

        Returns:
            UpdateResult with success/failure status and details
        """
        # Import version getter here to avoid circular imports
        from executor.version import get_version

        current_version = get_version()
        logger.info(f"Starting update check. Current version: {current_version}")

        # Check for updates
        print("Checking for updates...")
        if self.update_config.is_registry():
            registry_url = self.update_config.get_registry_url()
            logger.info(f"Update source: registry ({registry_url})")
            print(f"  Source: registry ({registry_url})")
        else:
            from executor.services.updater.github_version_checker import (
                GithubVersionChecker,
            )

            # Get the embedded or default GitHub repo
            github_repo = (
                GithubVersionChecker._EMBEDDED_GITHUB_REPO
                or GithubVersionChecker.DEFAULT_GITHUB_REPO
            )
            logger.info(f"Update source: GitHub ({github_repo})")
            print(f"  Source: GitHub ({github_repo})")
        update_info = await self.version_checker.check_for_updates(current_version)

        if update_info is None:
            # Either already on latest or API error
            logger.info(f"Already on latest version: {current_version}")
            return UpdateResult(
                success=True,
                already_latest=True,
                old_version=current_version,
            )

        # Update is available
        print(f"\nUpdate available: v{current_version} → v{update_info.version}")
        logger.info(f"Update available: {current_version} → {update_info.version}")
        logger.info(f"Download URL: {update_info.url}")

        # Check disk space before proceeding
        if not self._check_disk_space():
            logger.error("Insufficient disk space for update")
            return UpdateResult(
                success=False,
                error="Insufficient disk space (need ~150 MB free)",
                old_version=current_version,
                new_version=update_info.version,
            )

        # Prompt user for confirmation (unless auto_confirm is set)
        if not self.auto_confirm and not self._confirm_update():
            logger.info("Update cancelled by user")
            return UpdateResult(
                success=False,
                error="Update cancelled by user",
                old_version=current_version,
                new_version=update_info.version,
            )

        logger.info("User confirmed update, proceeding with download")

        print()

        # Determine auth token based on source
        if self.update_config.is_registry():
            auth_token = self.update_config.get_token()
        else:
            # GitHub assets don't need auth for public repos
            auth_token = None

        # Create binary replacer
        self.binary_replacer = BinaryReplacer(
            download_url=update_info.url,
            auth_token=auth_token,
        )

        try:
            # Download with progress
            print("Downloading update...")
            logger.info("Starting download...")
            new_binary = self.binary_replacer.download_binary(
                progress_callback=self._print_progress
            )
            print()  # New line after progress bar
            logger.info(f"Download complete: {new_binary}")

            # Get current binary path
            current_binary = self._get_current_binary_path()
            logger.info(f"Current binary path: {current_binary}")

            print("Installing update...")
            logger.info("Installing update...")

            # Replace binary
            success = self.binary_replacer.replace_binary(new_binary, current_binary)

            if not success:
                logger.error(
                    "Failed to replace binary (permission denied or file in use)"
                )
                return UpdateResult(
                    success=False,
                    error="Failed to replace binary (permission denied or file in use)",
                    old_version=current_version,
                    new_version=update_info.version,
                )

            logger.info(f"Update successful: {current_version} → {update_info.version}")
            return UpdateResult(
                success=True,
                old_version=current_version,
                new_version=update_info.version,
            )

        except RuntimeError as e:
            return UpdateResult(
                success=False,
                error=str(e),
                old_version=current_version,
                new_version=update_info.version,
            )
        except Exception as e:
            logger.exception("Unexpected error during update")
            return UpdateResult(
                success=False,
                error=f"Unexpected error: {e}",
                old_version=current_version,
                new_version=update_info.version,
            )

    def _get_current_binary_path(self) -> Path:
        """Get path to currently running binary.

        Returns:
            Path to the current executable

        Raises:
            RuntimeError: If not running from a PyInstaller binary
        """
        if not getattr(sys, "frozen", False):
            # Development mode - use the script path as fallback
            # This allows testing the update flow in development
            return Path(sys.argv[0]).resolve()

        return Path(sys.executable).resolve()

    def _confirm_update(self) -> bool:
        """Prompt user to confirm update.

        Returns:
            True if user confirms, False otherwise
        """
        try:
            response = input("Download and install update? [Y/n] ").strip().lower()
            return response in ("", "y", "yes")
        except (EOFError, KeyboardInterrupt):
            # Handle piped input or Ctrl+C
            return False

    def _check_disk_space(self) -> bool:
        """Check if there's enough free disk space.

        Returns:
            True if sufficient space available, False otherwise
        """
        try:
            home = Path.home()
            stat = shutil.disk_usage(home)
            if stat.free < self.MIN_FREE_SPACE:
                free_mb = stat.free // (1024 * 1024)
                required_mb = self.MIN_FREE_SPACE // (1024 * 1024)
                print(
                    f"✗ Insufficient disk space: {free_mb} MB free, {required_mb} MB required"
                )
                return False
            return True
        except Exception as e:
            logger.warning(f"Failed to check disk space: {e}")
            # Proceed anyway if we can't check
            return True

    def _print_progress(self, downloaded: int, total: Optional[int]) -> None:
        """Print download progress to terminal.

        Args:
            downloaded: Bytes downloaded so far
            total: Total bytes to download (None if unknown)
        """
        if self.binary_replacer is None:
            return

        progress = BinaryReplacer.format_progress_bar(downloaded, total)
        # Use carriage return to overwrite the same line
        print(f"\r{progress}", end="", flush=True)

File: executor/src/services/updater/updater_service.rs
// SPDX-FileCopyrightText: 2026 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

use std::{
    future::Future,
    path::{Path, PathBuf},
    pin::Pin,
};

use crate::{
    config::device::UpdateConfig,
    services::updater::{
        binary_name_for, BinaryReplacer, GithubVersionChecker, RegistryVersionChecker, UpdateInfo,
    },
    version::get_version,
};

pub const MIN_FREE_SPACE_BYTES: u64 = 150 * 1024 * 1024;

pub type BoxFutureResult<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;

pub trait UpdateChecker: Clone + Send + Sync + 'static {
    fn check_for_updates<'a>(
        &'a self,
        current_version: &'a str,
    ) -> BoxFutureResult<'a, Option<UpdateInfo>>;
}

pub trait BinaryInstaller: Clone + Send + Sync + 'static {
    fn install<'a>(
        &'a self,
        update: &'a UpdateInfo,
        current_binary: &'a Path,
        auth_token: Option<&'a str>,
    ) -> BoxFutureResult<'a, Result<(), String>>;
}

pub trait UpdateRuntime: Clone + Send + Sync + 'static {
    fn current_version(&self) -> String;
    fn current_binary_path(&self) -> PathBuf;
    fn has_sufficient_disk_space(&self) -> bool;
    fn confirm_update(&self) -> bool;
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct UpdateResult {
    pub success: bool,
    pub already_latest: bool,
    pub old_version: Option<String>,
    pub new_version: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone)]
pub struct UpdaterService<C, I, R>
where
    C: UpdateChecker,
    I: BinaryInstaller,
    R: UpdateRuntime,
{
    update_config: UpdateConfig,
    version_checker: C,
    installer: I,
    runtime: R,
    auto_confirm: bool,
}

impl<C, I, R> UpdaterService<C, I, R>
where
    C: UpdateChecker,
    I: BinaryInstaller,
    R: UpdateRuntime,
{
    pub fn with_components(
        update_config: UpdateConfig,
        version_checker: C,
        installer: I,
        runtime: R,
        auto_confirm: bool,
    ) -> Self {
        Self {
            update_config,
            version_checker,
            installer,
            runtime,
            auto_confirm,
        }
    }

    pub fn update_config(&self) -> &UpdateConfig {
        &self.update_config
    }

    pub fn auto_confirm(&self) -> bool {
        self.auto_confirm
    }

    pub async fn check_and_update(&self) -> UpdateResult {
        let current_version = self.runtime.current_version();
        let Some(update_info) = self
            .version_checker
            .check_for_updates(&current_version)
            .await
        else {
            return UpdateResult {
                success: true,
                already_latest: true,
                old_version: Some(current_version),
                ..UpdateResult::default()
            };
        };

        if !self.runtime.has_sufficient_disk_space() {
            return UpdateResult {
                success: false,
                old_version: Some(current_version),
                new_version: Some(update_info.version),
                error: Some("Insufficient disk space (need ~150 MB free)".to_owned()),
                ..UpdateResult::default()
            };
        }

        if !self.auto_confirm && !self.runtime.confirm_update() {
            return UpdateResult {
                success: false,
                old_version: Some(current_version),
                new_version: Some(update_info.version),
                error: Some("Update cancelled by user".to_owned()),
                ..UpdateResult::default()
            };
        }

        let current_binary = self.runtime.current_binary_path();
        match self
            .installer
            .install(&update_info, &current_binary, self.update_config.token())
            .await
        {
            Ok(()) => UpdateResult {
                success: true,
                old_version: Some(current_version),
                new_version: Some(update_info.version),
                ..UpdateResult::default()
            },
            Err(error) => UpdateResult {
                success: false,
                old_version: Some(current_version),
                new_version: Some(update_info.version),
                error: Some(error),
                ..UpdateResult::default()
            },
        }
    }
}

impl UpdaterService<DefaultUpdateChecker, DefaultBinaryInstaller, SystemUpdateRuntime> {
    pub fn new(update_config: UpdateConfig, auto_confirm: bool) -> Self {
        Self::with_components(
            update_config.clone(),
            DefaultUpdateChecker::from_config(&update_config),
            DefaultBinaryInstaller,
            SystemUpdateRuntime,
            auto_confirm,
        )
    }
}

#[derive(Debug, Clone)]
pub enum DefaultUpdateChecker {
    Github(GithubVersionChecker),
    Registry(RegistryVersionChecker),
}

impl DefaultUpdateChecker {
    pub fn from_config(config: &UpdateConfig) -> Self {
        if let Some(registry_url) = config.registry_url() {
            return Self::Registry(RegistryVersionChecker::new(registry_url, config.token()));
        }
        Self::Github(GithubVersionChecker::new(None))
    }
}

impl UpdateChecker for DefaultUpdateChecker {
    fn check_for_updates<'a>(
        &'a self,
        current_version: &'a str,
    ) -> BoxFutureResult<'a, Option<UpdateInfo>> {
        Box::pin(async move {
            let binary_name = binary_name_for(std::env::consts::OS, std::env::consts::ARCH);
            match self {
                Self::Github(checker) => {
                    checker
                        .check_for_updates_for_binary(current_version, &binary_name)
                        .await
                }
                Self::Registry(checker) => {
                    checker
                        .check_for_updates_for_binary(current_version, &binary_name)
                        .await
                }
            }
        })
    }
}

#[derive(Debug, Clone)]
pub struct DefaultBinaryInstaller;

impl BinaryInstaller for DefaultBinaryInstaller {
    fn install<'a>(
        &'a self,
        update: &'a UpdateInfo,
        current_binary: &'a Path,
        auth_token: Option<&'a str>,
    ) -> BoxFutureResult<'a, Result<(), String>> {
        Box::pin(async move {
            let replacer = BinaryReplacer::new(&update.download_url, auth_token);
            let new_binary = replacer
                .download_binary_to(&std::env::temp_dir(), |_downloaded, _total| {})
                .await
                .map_err(|error| error.to_string())?;
            if replacer.replace_binary(&new_binary, current_binary) {
                Ok(())
            } else {
                Err("Failed to replace binary (permission denied or file in use)".to_owned())
            }
        })
    }
}

#[derive(Debug, Clone)]
pub struct SystemUpdateRuntime;

impl UpdateRuntime for SystemUpdateRuntime {
    fn current_version(&self) -> String {
        get_version()
    }

    fn current_binary_path(&self) -> PathBuf {
        std::env::current_exe().unwrap_or_else(|_| PathBuf::from("wegent-executor"))
    }

    fn has_sufficient_disk_space(&self) -> bool {
        free_space_bytes(&home_dir())
            .map(has_sufficient_update_space)
            .unwrap_or(true)
    }

    fn confirm_update(&self) -> bool {
        true
    }
}

pub fn parse_update_confirmation(input: &str) -> bool {
    matches!(input.trim().to_ascii_lowercase().as_str(), "" | "y" | "yes")
}

pub fn has_sufficient_update_space(free_bytes: u64) -> bool {
    free_bytes >= MIN_FREE_SPACE_BYTES
}

#[cfg(unix)]
fn free_space_bytes(path: &Path) -> Option<u64> {
    use std::ffi::CString;
    use std::os::unix::ffi::OsStrExt;

    let c_path = CString::new(path.as_os_str().as_bytes()).ok()?;
    let mut stat = std::mem::MaybeUninit::<libc::statvfs>::uninit();
    let result = unsafe { libc::statvfs(c_path.as_ptr(), stat.as_mut_ptr()) };
    if result != 0 {
        return None;
    }
    let stat = unsafe { stat.assume_init() };
    let available_blocks = numeric_to_u64(stat.f_bavail)?;
    let fragment_size = numeric_to_u64(stat.f_frsize)?;
    Some(available_blocks.saturating_mul(fragment_size))
}

#[cfg(windows)]
fn free_space_bytes(path: &Path) -> Option<u64> {
    use std::os::windows::ffi::OsStrExt;
    use windows_sys::Win32::Storage::FileSystem::GetDiskFreeSpaceExW;

    let mut wide_path: Vec<u16> = path.as_os_str().encode_wide().collect();
    wide_path.push(0);
    let mut free_bytes_available = 0u64;
    let mut total_bytes = 0u64;
    let mut total_free_bytes = 0u64;
    let ok = unsafe {
        GetDiskFreeSpaceExW(
            wide_path.as_ptr(),
            &mut free_bytes_available,
            &mut total_bytes,
            &mut total_free_bytes,
        )
    };
    if ok == 0 {
        return None;
    }
    Some(free_bytes_available)
}

#[cfg(not(any(unix, windows)))]
fn free_space_bytes(_path: &Path) -> Option<u64> {
    None
}

fn home_dir() -> PathBuf {
    dirs::home_dir().unwrap_or_else(std::env::temp_dir)
}

#[cfg(unix)]
fn numeric_to_u64<T>(value: T) -> Option<u64>
where
    T: TryInto<u64>,
{
    value.try_into().ok()
}
